### PR TITLE
Add HashiCorp Nomad provider

### PR DESCRIPTION
**Pull request by anubhavmishra**
HashiCorp Nomad provider for Virtual Kubelet connects your Kubernetes cluster
with Nomad cluster by exposing the Nomad cluster as a node in Kubernetes. By
using the provider, pods that are scheduled on the virtual Nomad node
registered on Kubernetes will run as jobs on Nomad clients as they
would on a Kubernetes node.

TODO:
- [x] Figure out CI for Nomad tests.
- [x] vendor dependencies for Nomad API. 
- [x] Add snapshots and a gif to show a demo of the Nomad provider..

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -200,6 +201,20 @@ Service Fabric Mesh is a fully managed service that lets developers deploy micro
...
+with Nomad cluster by exposing the Nomad cluster as a node in Kubernetes. By
+using the provider, pods that are scheduled on the virtual Nomad node
+registered on Kubernetes will run as jobs on Nomad clients as they
+would on a Kubernetes node.
+
+```bash
+./bin/virtual-kubelet --provider="nomad"
+```
+
+For detailed instrustions, follow the guide [here](providers/nomad/README.md).
```
Comment (radu-matei) [inline README.md]: typo: _instrustions_

Comment (anubhavmishra) [reply README.md]: Thank you! Fixed. :)

File: .circleci/config.yml [hunk 2]
```diff
@@ -32,6 +32,16 @@ jobs:
           echo 'export AZURE_AUTH_LOCATION=${outputPathCredsfile}' >> $BASH_ENV
       - run: |
           echo 'export KUBECONFIG=${outputPathKubeConfigFile}' >> $BASH_ENV
+      - run: |
```
Comment (anubhavmishra) [inline .circleci/config.yml]: Not sure what is the best way to add the `nomad` binary to CircleCI. The binary is used for testing the provider. Just adding this as a step in CircleCI workflow.

File: providers/nomad/helpers.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+func jobStatusToPodPhase(status string) v1.PodPhase {
+	switch status {
+	case "pending":
+		return v1.PodPending
+	case "running":
+		return v1.PodRunning
+	// TODO: Make sure we take PodFailed into account.
+	case "dead":
```
Comment (robbiezhang) [inline providers/nomad/helpers.go]: just curious: do you have a "PodSucceeded" state?

File: .circleci/config.yml (hunk 2 above)
Comment (radu-matei) [reply .circleci/config.yml]: cc @robbiezhang

Comment (robbiezhang) [reply .circleci/config.yml]: give it a name "Install nomad"

Comment (robbiezhang) [reply .circleci/config.yml]: BTW, prefer to add it right before the "Tests" step

**Comment by anubhavmishra** [comment]
@robbiezhang `dep ensure` runs successfully without any changes to the project on my local machine. Here is the output:

```bash
$ dep ensure
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

File: providers/nomad/helpers.go (hunk 3 above)
Comment (anubhavmishra) [reply providers/nomad/helpers.go]: Currently, we don't have an equivalent state to "PodSucceeded". I will try taking a peak again if I could use any other state from the job spec.

File: providers/nomad/README.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    imagePullPolicy: Always
+    name: nginx
+    ports:
+    - containerPort: 80
+    - containerPort: 443
+  dnsPolicy: ClusterFirst
+  nodeSelector:
+    kubernetes.io/role: agent
+    beta.kubernetes.io/os: linux
+    type: virtual-kubelet
```
[the remark this answers is not in the document]
Comment (anubhavmishra) [reply providers/nomad/README.md]: Fixed.

**Comment by anubhavmishra** [comment]
Found the issue with `dep ensure` [here](https://github.com/golang/dep/issues/1994). It's all fixed and `dep check` works now.

**Comment by anubhavmishra** [comment]
recheck

**Comment by anubhavmishra** [comment]
@robbiezhang should be good to go! There is a transient test failures while running the aci but that isn't related to the Nomad provider. If we re-run the tests they will be green I am sure. 😄

**Comment by anubhavmishra** [comment]
Thanks @robbiezhang! All tests pass! 👍🏽

**Pull request merged**